pkg/apis/serving/v1beta1/inference_service_status.go

```diff
@@ -565,50 +566,73 @@ func (ss *InferenceServiceStatus) SetModelFailureInfo(info *FailureInfo) bool {
 	return true
 }
 
-func (ss *InferenceServiceStatus) PropagateModelStatus(statusSpec ComponentStatusSpec, podList *v1.PodList, rawDeployment bool) {
+func (ss *InferenceServiceStatus) PropagateModelStatus(statusSpec ComponentStatusSpec, podList *v1.PodList, rawDeployment bool, serviceStatus *knservingv1.ServiceStatus) bool {
 	// Check at least one pod is running for the latest revision of inferenceservice
 	totalCopies := len(podList.Items)
 	if totalCopies == 0 {
-		ss.UpdateModelRevisionStates(Pending, totalCopies, nil)
-		return
-	}
-	// Update model state to 'Loaded' if inferenceservice status is ready.
-	// For serverless deployment, the latest created revision and the latest ready revision should be equal
-	if ss.IsReady() {
-		if rawDeployment {
-			ss.UpdateModelRevisionStates(Loaded, totalCopies, nil)
-			return
-		} else if statusSpec.LatestCreatedRevision == statusSpec.LatestReadyRevision {
-			ss.UpdateModelRevisionStates(Loaded, totalCopies, nil)
-			return
+		if !rawDeployment {
+			// Make sure we haven't scaled down to 0 because of an error
+			for _, knativeCond := range serviceStatus.Conditions {
+				if knativeCond.Status == "False" {
+					// If any of the knative statuses are False, the model failed
+					// Hopefully the lastFailureInfo already has the info we need, so we don't update it here
+					ss.UpdateModelRevisionStates(FailedToLoad, totalCopies, nil)
+					return true
+				}
+			}
 		}
+
+		// If we made it here then hopefully there are 0 pods because we're just getting started and therefore Pending seems appropriate
+		ss.UpdateModelRevisionStates(Pending, totalCopies, nil)
+		return true
 	}
+
 	// Update model state to 'Loading' if storage initializer is running.
 	// If the storage initializer is terminated due to error or crashloopbackoff, update model
 	// state to 'ModelLoadFailed' with failure info.
 	for _, cs := range podList.Items[0].Status.InitContainerStatuses {
 		if cs.Name == constants.StorageInitializerContainerName {
 			switch {
 			case cs.State.Running != nil:
-				ss.UpdateModelRevisionStates(Loading, totalCopies, nil)
-				return
+				// Double check that we aren't missing an error
+				if cs.LastTerminationState.Terminated != nil {
```

Comment:
I am not sure if this extra check here has any effect.

Reply:
The intention was to only re-trigger the reconcile loop if there might be an issue (catch the case where the storage-initializer is looping between an error and running. I was seeing crashloop -> running -> terminated). But if there's no error then it would be a waste to re-check the statuses

Reply:
Ok, please add this explanation as comment.

Reply:
instead of triggering a new loop, what happens if you set the revision state?

Reply:
Without this change, the last time the reconcile loop happens it catches the storage-initializer in the running state and sets the revision state to Loading. 

I didn't directly just set the revision state to a Failure instead because I really wanted to catch the terminated state (that felt like the most correct sequence of events). The knative failure check is just a backup to make sure that if we still end up missing the terminated state that we don't see the same status issue that this PR is addressing (if we missed the terminated state and didn't check for knative failures we would end up as Pending because the deployment scaled down to 0).

Reply:
ok, make sense.